test/Benchmarks.cpp

```diff
@@ -53,6 +53,8 @@ static unsigned int currentColor = 0;
 using namespace care;
 using int_ptr = host_device_ptr<int>;
 
+#if defined(CARE_GPUCC)
+//This test is not valid on the CPU since AllocationStrategy chai::PINNED does not exist for that build.
```

Comment:
Is this true even in GPU sim mode?

Reply:
No. Are you thinking this should be `#if defined(CARE_GPUCC) || defined(CARE_ENABLE_GPU_SIMULATION_MODE)`?

Reply:
This and many other conditionals, I think.

Reply:
I tried recompiling as 

cmake .. -DCMAKE_CXX_COMPILER=/usr/tce/packages/intel/intel-19.0.4/bin/icpc -DCMAKE_BUILD_TYPE=Debug -DCARE_ENABLE_GPU_SIMULATION_MODE=ON

and I am still getting failures and segfaults in this test and other loop fuser tests, although sometimes in different places than before it seems. 

Previously, in LoopFuser.cpp everything was ifdefed out for non-GPU builds in favor of some dummy implementation, so I think this was not exercised even for GPU_SIMULATION

Reply:
ALE3D builds with the following:
# Make sure the resource manager is on
set(CHAI_DISABLE_RM "OFF" CACHE STRING "")
# Enable pinned memory
set(CHAI_ENABLE_PINNED "ON" CACHE STRING "")
# Ensure Loop Fuser is turned on 
set(CARE_ENABLE_LOOP_FUSER "ON" CACHE STRING "")

# Make sure lambdas are turned on
set(ENABLE_LEGACY_CARE "0" CACHE STRING "")

# Make sure GPU simulation mode is turned on
set(CHAI_ENABLE_GPU_SIMULATION_MODE "ON" CACHE STRING "")
set(CARE_ENABLE_GPU_SIMULATION_MODE "ON" CACHE STRING "")


Does this mean that the loop fuser is still disabled for that build?

Reply:
Just a sec, I'll do a test on the recent CARE develop version (not my branch) with your options and test what happens

Reply:
In my ALE3D repo, I checked out care and its dependencies, and in the file care/src/care/LoopFuser.cpp I made the following changes to see if it gets built:

// CARE config header
#include "care/config.h"

#if CARE_ENABLE_LOOP_FUSER

//see if this triggers an error on GPU SIM mode...
oogabooga!!!

// Other CARE headers
#include "care/DefaultMacros.h"


And then I did the following:

make pmake-all VARIANT=gpu_sim

It seems like it was being triggered. So it seems like the ale3d version was fine. I'll investigate a bit

Reply:
I'll take a closer look after lunch, sorry for the confusion.

Reply:
Good news! I compiled this with the development CARE version (NOT my branch) using the command  cmake .. -DDCMAKE_C_COMPILER=/usr/tce/packages/intel/intel-19.0.4/bin/icc -DCMAKE_CXX_COMPILER=/usr/tce/packages/intel/intel-19.0.4/bin/icpc -DCHAI_DISABLE_RM=OFF -DCHAI_ENABLE_PINNED=ON -DCARE_ENABLE_LOOP_FUSER=ON -DENABLE_LEGACY_CARE=0 -DCHAI_ENABLE_GPU_SIMULATION_MODE=ON -DCARE_ENABLE_GPU_SIMULATION_MODE=ON . Hopefully that is the right command?

I got the exact same errors as with my branch and with the development CARE branch. Happily, this means that any issues with the GPU SIM mode were not introduced as a result of this PR, but were per-exisiting.

Reply:
I did some more investigating. I built care as part of ale3d (so using the version of care that is currently used in ale3d), building in gpu-sim mode and turning tests on (they were off by default).

It seems that the Benchmarks test is still failing in the version of CARE used by ale3d in GPU SIM mode - more evidence that this was a pre-existing issue not caused by this PR.

Reply:
Ok, so the tests for GPU_SIM mode will need to be revisited elsewhere.

Reply:
There's a broader conversation here - I think it might be reasonable for care::initialize_pool to contain the preprocesor logic to convert things that aren't enabled into fallback options, but I could be convinced otherwise.